ci: Treat IWYU violations as errors

hebasto opened the pull request:
Fixes the https://github.com/bitcoin/bitcoin/pull/24831#discussion_r861835262:
> Not sure if this is setting the return code to non-zero if it fails?

MarcoFalke commented on src/util/string.cpp:
```diff
@@ -4,6 +4,7 @@
 
 #include <util/string.h>
 
+#include <map>
```
This is wrong. Maybe use libc++ instead of libstdc++? Though, that will probably be wrong as well, see https://github.com/include-what-you-use/include-what-you-use/issues/1110#issuecomment-1245370908

hebasto replied:
What about
```diff
--- a/src/util/string.cpp
+++ b/src/util/string.cpp
@@ -11,5 +11,5 @@
 void ReplaceAll(std::string& in_out, const std::string& search, const std::string& substitute)
 {
     if (search.empty()) return;
-    in_out = std::regex_replace(in_out, std::regex(search), substitute);
+    in_out = std::regex_replace(in_out, std::regex(search.data()), substitute);
 }
```
?

Looking into `bits/regex.h` makes me believe it won't introduce any regressions.

fanquake replied:
There is a problem if we are refactoring code for no other reason than to make a buggy tool happy.

hebasto replied:
https://github.com/bitcoin/bitcoin/blob/682a87e0db1fbddf20c40ab6bce9c70cad62e87a/src/util/string.cpp#L9-L10

fanquake commented on src/primitives/transaction.h:
```diff
@@ -22,6 +22,9 @@
 #include <tuple>
 #include <utility>
 #include <vector>
+#if __cplusplus >= 202002L
```
Not a fan of adding stuff like this just to appease a tool, which happens to be getting run under a non-default compilation mode (`--enable-c++20`).

hebasto replied:
See 89bcb1a856a563013062c2c129012498e6b8b1f9

fanquake replied:
I don't think we want to just drop the option without adding it to a different CI (and not a Windows build).

hebasto replied:
> I don't think we want to just drop the option without adding it to a different CI (and not a Windows build).

Agree. Also https://github.com/bitcoin/bitcoin/pull/25528#discussion_r1058955188

Getting ready a proposal.

hebasto replied:
> Getting ready a proposal.

#26770

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/26763#issuecomment-1367571160).

fanquake commented on src/compat/compat.h:
```diff
@@ -25,10 +25,10 @@
 #include <sys/mman.h>
 #include <sys/select.h>
-#include <sys/socket.h>
+#include <sys/socket.h> // IWYU pragma: export
```
What symbols are these needed for (here and below), and what are the trade-offs between adding the pragma, and actually adding the include where it is used? We should probably have some rationale as to wether we are going to further include/"export" everything out of compat, or actually add includes to source files.

hebasto replied:
> What symbols are these needed for (here and below)...

```
#include <sys/socket.h>                    // for socklen_t, size_t, ssize_t
```
```
#include <netinet/in.h>                    // for in6_addr, in_addr, s6_addr
```

Both headers are non-Windows.

> ... and what are the trade-offs between adding the pragma, and actually adding the include where it is used?

Platform-specific headers require to be guarded with `#ifdef ... #endif` macros. So adding them in every place, where they are required, is not very [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself).

> We should probably have some rationale as to wether we are going to further include/"export" everything out of compat, or actually add includes to source files.

My suggestion is to "export" platform-specific headers out of `compat.h` when IWYU points at them.

fanquake replied:
> #include <sys/socket.h> .... size_t, ssize_t

Is this a bug? `size_t`/`ssize_t` do not come from `socket.h`.

> So adding them in every place, where they are required, is not very [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself).

Isn't adding every header, everywhere it is used, the entire point of IWYU though?

This could also lead to having a single bloated compat.h, which when included everywhere, mitigates one of the other supposed benefit of IWYU, which is removing unneeded headers to reducing preprocessing time.

MarcoFalke replied:
It just seems more convenient to include compat.h or socks.h when using the `Sock` wrapper, instead of having to deal with low-level system includes everywhere `Sock` is used. I guess an alternative to the iwyu pragma would be to also wrap the low-level types, but not sure if this is actually better.

hebasto replied:
> > #include <sys/socket.h> .... size_t, ssize_t
> 
> Is this a bug? `size_t`/`ssize_t` do not come from `socket.h`.

Perhaps, it is. But it is unrelated, as the `socket.h` header is required to provide the `socklen_t` symbol.

> > So adding them in every place, where they are required, is not very [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself).
> 
> Isn't adding every header, everywhere it is used, the entire point of IWYU though?
> 
> This could also lead to having a single bloated compat.h, which when included everywhere, mitigates one of the other supposed benefit of IWYU, which is removing unneeded headers to reducing preprocessing time.

That is not what I meant. The `compat.h` header, which includes _platform-specific_ headers, allows to avoid repetition of code like that:
```c++
#ifdef WIN32
#include <winsock2.h>
#else
#include <sys/socket.h>
#endif
```

hebasto commented:
Rebased 682a87e0db1fbddf20c40ab6bce9c70cad62e87a -> 9b7443dc27ac9b3188a2a8ace51478970b98b3b7 ([pr26763.02](https://github.com/hebasto/bitcoin/commits/pr26763.02) -> [pr26763.03](https://github.com/hebasto/bitcoin/commits/pr26763.03))  on top of the bitcoin/bitcoin#26770.

ajtowns commented on src/chain.h:
```diff
@@ -7,15 +7,24 @@
 #include <vector>
 
+namespace Consensus {
+struct Params;
```
Why drop the `#include <consensus/params.h>` just to manually forward declare the struct? This seems like a backwards step. Seems better to either leave it as is, or move `GetBlockProofEquivalentTime` back to `pow.cpp` so that `chain.h` doesn't need `Consensus::Params` at all (partially reverting #7311 ; GBPET() is used in ConnectBlock() so it isn't really separate from consensus functionality for us).

hebasto replied:
> Why drop the `#include <consensus/params.h>` just to manually forward declare the struct?

Although forward declarations have their own [pros and cons](https://google.github.io/styleguide/cppguide.html#Forward_Declarations), this code base seems to lean in their favour.

> This seems like a backwards step.

Mind elaborating?

---

As a question like that has been raised, should we document our preferences in https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md?

ajtowns replied:
> Mind elaborating?

If we've got a dependency from one module to another ("validation" depends on "consensus/params") then we should just include the header file; if we're avoiding a dependency (eg making "chain" just store chain data independent of whatever the consensus parameters might be), then we shouldn't be entangling them by referencing the "unrelated" module's definitions.

Some of those cases don't even avoid the dependency at all; eg validation.h includes chainparams.h which includes consensus/params.h anyway.

It makes the dependencies slightly harder to track: forward declarations are harder to search for than a plain #include; it also violates the DRY-principle, though not in a very big way.

The benefit is that not including the full header can reduce compilation time in the cases where you can avoid including a complicated header (eg, one that pulls in boost multiindex or that pulls in lots of other headers). That probably makes sense if you're avoiding pulling in txmempool.h (which validation.h does), but even then any effort's probably better spent in improving our modularisation...

hebasto replied:
The changes in `src/chain.h` appeared non-required for this PR, and they have been [dropped](https://github.com/bitcoin/bitcoin/pull/26763#issuecomment-1370769788).

The discussion about forward declaration vs header including could be continued elsewhere, maybe in a dedicated issue, couldn't it?

ajtowns commented on src/versionbits.h:
```diff
@@ -6,6 +6,7 @@
 #define BITCOIN_VERSIONBITS_H
 
 #include <chain.h>
+#include <consensus/params.h>
```
I'm surprised `deploymentstatus.h` doesn't also need to include consensus/params.h, given it defines (non-template) inline functions that call `Consensus::ValidDeployment()`

hebasto replied:
According to https://api.cirrus-ci.com/v1/task/5220634955874304/logs/ci.log, the `deploymentstatus.h` is not parsed by IWYU now.

ajtowns reviewed (commented):
Concept ACK

hebasto commented:
Updated 9b7443dc27ac9b3188a2a8ace51478970b98b3b7 -> a6979504e36d75bfe68215b99d8bd7651843db53 ([pr26763.03](https://github.com/hebasto/bitcoin/commits/pr26763.03) -> [pr26763.04](https://github.com/hebasto/bitcoin/commits/pr26763.04)):

- rebased
- dropped non-required changes

MarcoFalke commented on src/test/fuzz/util/net.cpp:
```diff
@@ -21,6 +21,7 @@
 #include <cstdlib>
 #include <cstring>
 #include <thread>
+#include <unordered_map>
```
Looks wrong as well?

hebasto replied:
Well, in `FuzzedSock::WaitMany()` function https://github.com/bitcoin/bitcoin/blob/4717a5aa31bc9cdb5406f4161f4c600e5c2659c0/src/test/fuzz/util/net.cpp#L332-L339 the type `EventsPerSock` is actually an `std::unordered_map`: https://github.com/bitcoin/bitcoin/blob/2ec97825e70882262ee2b7d53dc1def788a60162/src/util/sock.h#L218

MarcoFalke replied:
Right, so it would need to include `util/sock.h` instead to get `EventsPerSock`?

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/26763#issuecomment-1370913829).

Some relevant IWYU's docs: https://github.com/include-what-you-use/include-what-you-use/blob/master/docs/WhatIsAUse.md#re-exporting.

MarcoFalke replied:
This will force iwyu to remove `#include <unordered_map>` if `util/sock.h` is included, but an unrelated code piece needs the unordered map include?

Not sure which is preferable. I am just explaining the tradeoffs.

hebasto replied:
What about `// IWYU pragma: no_include <unordered_map>` in `src/test/fuzz/util/net.cpp`?

MarcoFalke replied:
Seems worse (combining all downsides), compared to just including the file

hebasto replied:
FWIW, I lean to just following IWYU's suggestion and including a header. It won't make a translation unit bigger anyway.

hebasto commented:
Updated a6979504e36d75bfe68215b99d8bd7651843db53 -> ba0780b756d446fbdcb2102f0b39f4bf516ffe74 ([pr26763.04](https://github.com/hebasto/bitcoin/commits/pr26763.04) -> [pr26763.05](https://github.com/hebasto/bitcoin/commits/pr26763.05), [diff](https://github.com/hebasto/bitcoin/compare/pr26763.04..pr26763.05)):

- addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/26763#discussion_r1061370409)

MarcoFalke commented:
Needs rebase

hebasto commented:
> Needs rebase

Rebased ba0780b756d446fbdcb2102f0b39f4bf516ffe74 -> 077376c805c620ddb1a893bff4cdabc51ef44313 ([pr26763.05](https://github.com/hebasto/bitcoin/commits/pr26763.05) -> [pr26763.06](https://github.com/hebasto/bitcoin/commits/pr26763.06)).

MarcoFalke commented:
Concept ACK. Seems like a good attempt to reduce review on the include list or avoid shipping a release with missing includes (on Linux fwiw).

However, the CI output might be too cumbersome to act on. It would be nice if it also printed a nice diff to copy (or if there was a documented way to create one locally).

fanquake commented:
> However, the CI output might be too cumbersome to act on. It would be nice if it also printed a nice diff to copy (or if there was a documented way to create one locally).

Yea. I think in it's current state, this change will mostly just annoy contributors, for not much gain. Even more so if the CI is giving them nonsensical output (IWYU bugs), and it's not clear what should be done (add the wrong include, add a IWYU pragma inline, add a workaround to our `.imp` file etc).

hebasto commented:
> However, the CI output might be too cumbersome to act on. It would be nice if it also printed a nice diff to copy (or if there was a documented way to create one locally).

It is possible to leverage the `fix_includes.py` tool which produces [output](https://api.cirrus-ci.com/v1/task/6576481640382464/logs/ci.log) as follows:
```
>>> Fixing #includes in 'dbwrapper.h'
@@ -11,20 +11,19 @@
 #include <serialize.h>
 #include <span.h>
 #include <streams.h>
-
 #include <leveldb/db.h>
 #include <leveldb/iterator.h>
 #include <leveldb/options.h>
 #include <leveldb/slice.h>
 #include <leveldb/status.h>
 #include <leveldb/write_batch.h>
-
 #include <cstddef>
 #include <cstdint>
 #include <exception>
 #include <stdexcept>
 #include <string>
 #include <vector>
+#include <optional>

 namespace leveldb {
 class Env;
>>> Fixing #includes in 'kernel/context.cpp'
@@ -3,13 +3,10 @@
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.

 #include <kernel/context.h>
-
 #include <crypto/sha256.h>
 #include <key.h>
 #include <logging.h>
-#include <pubkey.h>
 #include <random.h>
-
 #include <string>


IWYU edited 2 files on your behalf.
```

---

> I think in it's current state, this change will mostly just annoy contributors, for not much gain. Even more so if the CI is giving them nonsensical output (IWYU bugs), and it's not clear what should be done (add the wrong include, add a IWYU pragma inline, add a workaround to our `.imp` file etc).

The current state, in the master branch, is worse because using IWYU is a no-op effectively, but consumes resources.

In case IWYU produces "nonsensical output" for any particular source file or header, the latter could just be skipped from analysis temporarily.

MarcoFalke commented:
Looks like that would need `--blank_lines` set to leave the new lines? Also, it would be good to have a single patch to copy-paste for all files, and not patches for each file separated by debug logs.

fanquake commented:
> The current state, in the master branch, is worse because using IWYU is a no-op effectively, but consumes resources.

I disagree, it's not worse, because if you don't care about it, you don't have to engage / fix anything, and the resource usage/overhead is basically 0. It's also useful enough that if someone does want to use the tool / run it locally, they can.

hebasto commented:
> Looks like that would need `--blank_lines` set to leave the new lines?

It doesn't help, unfortunately.

hebasto commented:
Rebased and drafted.

MarcoFalke commented:
> It is possible to leverage the fix_includes.py tool 

Could probably make sense to add this to the CI, to allow easy copy-pasting from Cirrus or from a local run.

jonatack commented on ci/test/06_script_b.sh:
```diff
@@ -72,7 +72,7 @@ if [ "${RUN_TIDY}" = "true" ]; then
           " src/util/threadinterrupt.cpp"\
           " src/zmq"\
-          " -p . ${MAKEJOBS} -- -Xiwyu --cxx17ns -Xiwyu --mapping_file=${BASE_BUILD_DIR}/bitcoin-$HOST/contrib/devtools/iwyu/bitcoin.core.imp"
+          " -p . ${MAKEJOBS} -- -Xiwyu --error -Xiwyu --cxx17ns -Xiwyu --mapping_file=${BASE_BUILD_DIR}/bitcoin-$HOST/contrib/devtools/iwyu/bitcoin.core.imp"
```
Light concept ACK if this change reduces review work and unburdens the review process, which seems more important than the potential annoyance it may add to opening a PR.

MarcoFalke commented:
CI fails with 


```
should remove these lines:
- #include <threadsafety.h>  // lines 9-9

hebasto committed:
iwyu: Fix IWYU violations

hebasto committed:
ci: Treat IWYU violations as errors

hebasto commented:
> CI fails

Thanks! Fixed.

hebasto commented:
See https://github.com/bitcoin/bitcoin/pull/31308.
